### PR TITLE
CUDA GreedySearch ProcessLogits optimization

### PR DESCRIPTION
**Pull request by hariharans29**
### Description
Explore the possible re-use of the logits buffer in `GreedySearch` for cases where sequence length == 1 (Post the first decoding run, the sequence length is guaranteed to be 1). This re-use will ensure that we do not have to make copies of the logits before processing them. Currently, we make a copy of the logits even if the sequence length == 1 which is not necessary as we can directly re-use the logits buffer for the token generation step. A similar optimization exists in `BeamSearch`, but seems lacking in `GreedySearch`. Since, the logits buffer may contain padded data, we need to adjust the pieces consuming the logits buffer directly to account for any padding.

A more invasive change (needs changes in a few places) will be to adjust the interfaces of `ProcessLogits()`  such that it takes a reference to the logits and not a const reference as (based on my understanding) this is the only place where the logits from the decoder subgraph will ever be used and giving the `ProcessLogits()` method license to mutate/process the underlying buffer of the logits OrtValue seems reasonable (instead of making a copy and then mutating/processing them). The will also remove the ugly `const_cast`(s) seen in this change.

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/transformers/beam_search_impl.cu [hunk 1]
```diff
@@ -75,66 +76,69 @@ __global__ void LogitsProcessKernel(
...
-    // RepetitionPenaltyLogitsProcessor
-    if (repetition_penalty != 1.0f) {
-      int32_t* current_sequence = sequences + batch_beam_index * max_sequence_length;
-      bool found = false;
-      for (int i = 0; i < current_sequence_length; i++) {
-        if (current_sequence[i] == word_id) {
-          found = true;
-          break;
+    // Only process within the padding region
+    if (word_id < vocab_size) {
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/transformers/beam_search_impl.cu:83]: Only change in LogitsProcesskernel() are from lines 79 - 83

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc [hunk 2]
```diff
@@ -527,12 +542,14 @@ Status GreedySearchProcessLogits(
   }
 
   cuda::LaunchLogitsProcessKernel<CudaT>(
-      reinterpret_cast<CudaT*>(next_token_scores.data()),
+      is_reuse_logits_buffer ? const_cast<CudaT*>(logits_data)
```
Comment (wangyems) [inline onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:545]: maybe add a todo here for potential interface changes in ProcessLogits() in case we get a chance in future

Comment (hariharans29) [reply onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:545]: Need to add some comments in some other places too. Will do.

File: onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc [hunk 3]
```diff
@@ -526,13 +542,19 @@ Status GreedySearchProcessLogits(
                                          cudaMemcpyHostToDevice, cuda_stream));
   }
 
+  // TODO(hasesh): Can we avoid the const_cast by changing the interface of
+  // GreedySearchProcessLogits() to take in a non-const OrtValue for logits
+  // as this is the only place we will ever use the logits and it may be reasonable
+  // to allow this method to mutate/process the logits in-place
   cuda::LaunchLogitsProcessKernel<CudaT>(
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:549]: > cuda::LaunchLogitsProcessKernel<CudaT>( [](http://example.com/codeflow?start=1&length=40)

We can set the logits value in range [vocab_size, padded_vocab_size] to smallest, and then you don't need to change the TopK

Comment (hariharans29) [reply onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:549]: In LaunchLogitsProcessKernel() ?

Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.cc:549]: Yes

File: onnxruntime/contrib_ops/cuda/transformers/beam_search_impl.cu [hunk 4]
```diff
@@ -75,66 +76,71 @@ __global__ void LogitsProcessKernel(
...
-    if (repetition_penalty != 1.0f) {
-      int32_t* current_sequence = sequences + batch_beam_index * max_sequence_length;
-      bool found = false;
-      for (int i = 0; i < current_sequence_length; i++) {
-        if (current_sequence[i] == word_id) {
-          found = true;
-          break;
+    if (word_id >= vocab_size) {
+      // Set any value within the padding region to the lowest value so that it isn't picked
+      next_token_scores[index] = cub::FpLimits<T>::Lowest();
```
Comment (hariharans29) [inline onnxruntime/contrib_ops/cuda/transformers/beam_search_impl.cu:84]: Don't want to consider these tokens as they are just padding

**Review by yufenglee — APPROVED** [review]
:shipit:

**Review by wangyems — APPROVED** [review]
